controllers/containerimagebuild_resources.go

```diff
@@ -0,0 +1,368 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	policyv1beta1 "k8s.io/api/policy/v1beta1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	forgev1alpha1 "github.com/dominodatalab/forge/api/v1alpha1"
+)
+
+// creates all supporting resources required by build job
+func (r *ContainerImageBuildReconciler) checkPrerequisites(ctx context.Context, cib *forgev1alpha1.ContainerImageBuild) error {
+	if err := r.checkServiceAccount(ctx, cib); err != nil {
+		return err
+	}
+	if err := r.checkPodSecurityPolicy(ctx, cib); err != nil {
+		return err
+	}
+	if err := r.checkRole(ctx, cib); err != nil {
+		return err
+	}
+	if err := r.checkRoleBinding(ctx, cib); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// creates build job service account when missing
+func (r *ContainerImageBuildReconciler) checkServiceAccount(ctx context.Context, cib *forgev1alpha1.ContainerImageBuild) error {
+	return r.withOwnedResource(ctx, cib, &corev1.ServiceAccount{}, func() interface{} {
+		return &corev1.ServiceAccount{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      cib.Name,
+				Namespace: cib.Namespace,
+				Labels:    cib.Labels,
+			},
+		}
+	})
+}
+
+// creates build job pod security policy when missing
+func (r *ContainerImageBuildReconciler) checkPodSecurityPolicy(ctx context.Context, cib *forgev1alpha1.ContainerImageBuild) error {
+	return r.withOwnedResource(ctx, cib, &policyv1beta1.PodSecurityPolicy{}, func() interface{} {
+		psp := &policyv1beta1.PodSecurityPolicy{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      cib.Name,
+				Namespace: cib.Namespace,
+				Labels:    cib.Labels,
+				Annotations: map[string]string{
+					"seccomp.security.alpha.kubernetes.io/allowedProfileNames": "unconfined,runtime/default",
+					"apparmor.security.beta.kubernetes.io/allowedProfileNames": "unconfined,runtime/default",
+					"seccomp.security.alpha.kubernetes.io/defaultProfileName":  "unconfined",
+					"apparmor.security.beta.kubernetes.io/defaultProfileName":  "unconfined",
+				},
+			},
+			Spec: policyv1beta1.PodSecurityPolicySpec{
+				AllowPrivilegeEscalation: pointer.BoolPtr(true),
```

Comment:
Should this be gated by `GrantFullPrivilege` too?

Reply:
Good question. The answer is "no" because the standard workflow (non-root) requires privilege escalation in order to umask/rexec so we want to set this regardless of the mode.